shared/granularity_specific_io.go

```diff
@@ -0,0 +1,89 @@
+package shared
+
+import (
+	"github.com/spacemeshos/bitstream"
+	"io"
+)
+
+// GranSpecificReader provides a wrapper for io.Reader to allow granularity-specific
+// access to the stream according to the defined item size, where bit-granular and
+// byte-granular sizes are supported via a specialized code path.
+type GranSpecificReader struct {
+	ReadNext       func() ([]byte, error)
+	ReadNextUintBE func() (uint64, error)
+}
+
+func NewGranSpecificReader(rd io.Reader, itemBitSize uint) *GranSpecificReader {
+	gsReader := new(GranSpecificReader)
+	if itemBitSize%8 == 0 {
+		// Byte-granular reader is using the underlying reader directly.
+		gsReader.ReadNext = func() ([]byte, error) {
+			b := make([]byte, itemBitSize/8)
+			_, err := io.ReadFull(rd, b)
+			if err != nil {
+				return nil, err
+			}
+			return b, nil
+		}
+		gsReader.ReadNextUintBE = func() (uint64, error) {
+			b, err := gsReader.ReadNext()
+			if err != nil {
+				return 0, err
+			}
+			return UintBE(b), nil
+		}
+	} else {
+		// Bit-granular reader is using bitstream as a wrapper for the underlying reader.
+		br := bitstream.NewReader(rd)
+		gsReader.ReadNext = func() ([]byte, error) {
+			return br.Read(itemBitSize)
+		}
+		gsReader.ReadNextUintBE = func() (uint64, error) {
+			return br.ReadUint64BE(int(itemBitSize))
+		}
+	}
```

Comment:
As discussed, we would like to keep a specialized code path for these 2 cases at this layer, since the third-party lib optimization isn’t guaranteed.